### PR TITLE
fixes #8690 - converting repository to scoped search

### PR DESCRIPTION
**Pull request by jlsherrill**

### CODE REVIEW

**Comment by jlsherrill** [comment]
Note, as part of this PR, i convert the content view 'add repositories' page to use a param instead of search.  After playing around, search did not seem to work well for this use case.  So here i present two variations to solving the issue:

1) Simply adds an available_to_content_view param:

https://github.com/jlsherrill/katello/commit/51165a3896e68a2edf329eae824b1242f1efd69c#diff-4a58e16252386909489b00a2f4451879R67

2)  A bit more advanced adds an 'available_for' param that when used, and set to the value of 'content_view' changes the meaning of the content_view_id param, from one of 'what repos are associated with' to one of 'what repos are available to be associated with'

https://github.com/Katello/katello/commit/06f134c3550069690a63b9355bd66a4418966867#diff-4a58e16252386909489b00a2f4451879R62

**Comment by ehelms** [comment]
On Thu, Jun 11, 2015 at 11:17 PM, Justin Sherrill <notifications@github.com>
wrote:

> Note, as part of this PR, i convert the content view 'add repositories'
> page to use a param instead of search. After playing around, search did not
> seem to work well for this use case. So here i present two variations to
> solving the issue:
>
> 1) Simply adds an available_to_content_view param:
>
> jlsherrill@51165a3#diff-4a58e16252386909489b00a2f4451879R67
> <https://github.com/jlsherrill/katello/commit/51165a3896e68a2edf329eae824b1242f1efd69c#diff-4a58e16252386909489b00a2f4451879R67>
>

What about available_for_content_view? Reads better to me.


> 2) A bit more advanced adds an 'available_for' param that when used, and
> set to the value of 'content_view' changes the meaning of the
> content_view_id param, from one of 'what repos are associated with' to one
> of 'what repos are available to be associated with'
>
> 06f134c#diff-4a58e16252386909489b00a2f4451879R62
> <https://github.com/Katello/katello/commit/06f134c3550069690a63b9355bd66a4418966867#diff-4a58e16252386909489b00a2f4451879R62>
>

What are your thoughts on the second approach being more automatic? For
example, `available_for=content_view_id` where the code takes whatever
parameter is passed in for available_for, looks for that parameter in the
params and then performs the 'negation'. I realize this still couldn't work
for some parameters, and would only allow negation of a single parameter at
a time; however, without that I am not sure I see an advantage over method
1 that provides an explicit parameter. The plus side to this approach would
be you can take an existing query, and add a parameter to negate instead of
having to craft the query fresh.

**Comment by jlsherrill** [comment]
The negation is a lot trickier than what you'd expect.  Take this example from the repos controller:

```
     elsif params[:content_view_id]
        query = query.joins(:content_view_repositories).where("#{ContentViewRepository.table_name}.content_view_id" => params[:content_view_id])
      end
```

To negate that via sql-logic, you'd say something like:
```
     elsif params[:content_view_id]
        query = query.joins(:content_view_repositories).where("#{ContentViewRepository.table_name}.content_view_id != #{params[:content_view_id]}")
      end
```

You'd expect that would return you the opposite of the first query, but in fact it does not simply because the join() here is an inner join and excludes all things that are not associated to a content view at all.  Sure you could re-write this query to use an inner join and negate that, but what you are suggesting would require rewriting most all of our filter queries to use 'in clauses', which might be an option if it not for....

Available repos for a content view isn't simply the inverse of repos already associated to the content view.  You have to consider org, only the repos in Library (default org view), etc...  (notice here i set ```params[:library] =  true```, otherwise i'd get the literal inverse which would be wayyy too much.)  In most other situations the same will happen, available_for is not simply the inverse of the initial query as there are other things control for.

**Comment by jlsherrill** [comment]
Yeah, i agree, if we go that route, i'd change it to 'available_for_content_view', i started using available_for for the 2nd route cause it sounded better and just didn't go back and change the initial param

**Comment by ehelms** [comment]
I see your point. Thoughts on trying to capture the parameter logic into a function similar to how the content concern works? Something like:

```
filter_by_content_view(params[:content_view_id], params[:available_for] == 'content_view')
```

**Comment by jlsherrill** [comment]
@ehelms updated and added a test

**Comment by ehelms** [comment]
Probably not a symptom of this change, but when listing docker repositories for a content view version I see not just the library version listed.

**Comment by ehelms** [comment]
Do you think we should change the repository list page to use normal search instead of filter with this update?

**Comment by jlsherrill** [comment]
@ehelms i do think we should add that, but it would be best to just re-write that page to use the details-nutupane template, rather than modifying the existing code and i'd rather do that in a separate PR

**Comment by ehelms** [comment]
I thought with past conversions we had updated the pages to take advantage of the conversion, if you feel strongly enough in this case that is should be a stand-alone update please file an issue.

**Comment by jlsherrill** [comment]
@ehelms generally they are already using nutupane or nutupane-details and no conversion has been needed.  I started to do the conversion but with all the interaction and given the size of this PR already I didn't feel comfortable doing it as part of this PR.  

Opened http://projects.theforeman.org/issues/10837

**Comment by jlsherrill** [comment]
squashed

File: app/controllers/katello/api/v2/repositories_controller.rb [hunk 1]
```diff
@@ -35,56 +37,49 @@ class Api::V2::RepositoriesController < Api::V2::ApiController
     param :product_id, :number, :desc => N_("ID of a product to show repositories of")
     param :environment_id, :number, :desc => N_("ID of an environment to show repositories in")
     param :content_view_id, :number, :desc => N_("ID of a content view to show repositories in")
+    param :available_to_content_view_id, :number, :desc => N_("ID of content view to show repositories not in")
```
Comment (ehelms) [inline app/controllers/katello/api/v2/repositories_controller.rb]: Needs removing

File: app/controllers/katello/api/v2/repositories_controller.rb [hunk 2]
```diff
@@ -35,56 +37,49 @@ class Api::V2::RepositoriesController < Api::V2::ApiController
...
-      end
+      options = {:includes => [:gpg_key, :product, :environment]}
+      respond(:collection => scoped_search(index_relation.uniq, :name, :desc, options))
+    end
 
-      repositories = repositories.where(:content_type => params[:content_type]) if params[:content_type]
-      repositories = repositories.where(:name => params[:name]) if params[:name]
-      repositories = repositories.joins(:errata).where("#{Erratum.table_name}.uuid" => params[:erratum_id]) if params[:erratum_id]
+    def index_relation
+      query = Repository.where(:product_id => Product.readable.where(:organization_id => @organization.id))
```
Comment (ehelms) [inline app/controllers/katello/api/v2/repositories_controller.rb]: I know this is here from before, but we are saying if you can't view products (but could say view content views), you can't view repositories? I noticed in the UI, for content types, we check view on products or content views.

File: app/controllers/katello/api/v2/repositories_controller.rb [hunk 3]
```diff
@@ -35,56 +37,49 @@ class Api::V2::RepositoriesController < Api::V2::ApiController
...
+        query = filter_by_content_view(query, params[:content_view_id], params[:environment_id], params[:available_for] == 'content_view')
+      end
 
-        if params[:library]
-          repositories = Repository.where(:id => repositories.map(&:library_instance_id))
-        end
+      if params[:content_view_version_id]
+        query = query.where(:content_view_version_id => params[:content_view_version_id])
+        query = Repository.where(:id => query.map(&:library_instance_id)) if params[:library]
       end
```
Comment (ehelms) [inline app/controllers/katello/api/v2/repositories_controller.rb]: Seems like if I provide a content_view_version_id and a content_view_id then I wouldn't need to perform the `filter_by_content_view` method?

File: app/models/katello/repository.rb [hunk 4]
```diff
@@ -92,9 +89,13 @@ class Repository < Katello::Model
     scope :non_puppet, where("content_type != ?", PUPPET_TYPE)
     scope :non_archived, where('environment_id is not NULL')
     scope :archived, where('environment_id is NULL')
-
     scope :completer_scope, lambda { |options| options[table_name].call(self) }
 
+    scoped_search :on => :name, :complete_value => true
+    scoped_search :rename => :product, :on => :name, :through => :product, :complete_value => true
+    scoped_search :on => :content_type, :complete_value => Katello::Repository::TYPES.each_with_object({})  { |value, hash| hash[value.to_sym] = value }
+    scoped_search :on => :content_view_id, :in => :content_view_repositories
```
Comment (ehelms) [inline app/models/katello/repository.rb]: Can this be done via name similar to product?

Comment (jlsherrill) [inline app/models/katello/repository.rb]: in what way do you mean?

Comment (ehelms) [inline app/models/katello/repository.rb]: Search by content view name

File: app/controllers/katello/api/v2/repositories_controller.rb (hunk 3 above)
Comment (jlsherrill) [inline app/controllers/katello/api/v2/repositories_controller.rb]: true, but what if you provide 
applicable_for=content_view&content_view_id=5&content_view_version_id=3
or:
content_view_id=5&content_view_version_id=3
where version_id 3 isn't even in content_view_id 5.   

I'm not sure it makes sense to try to ignore some options in the face of other options, I'd just let the queries do what they do and the user will get no results if the query makes no sense

**Comment by ehelms** [comment]
I think some of the work here, the UI portion (https://github.com/Katello/katello/pull/5268/files), may have some coincidence with this update. You may want to review it, and perhaps consider getting it merged and then rebasing this PR as the autocomplete nature would be able to be added fairly easily?

File: app/models/katello/repository.rb (hunk 4 above)
Comment (jlsherrill) [inline app/models/katello/repository.rb]: actually this isn't currently possible due to https://github.com/wvanbergen/scoped_search/issues/131

**Comment by jlsherrill** [comment]
@ehelms testing that patch, but i don't know that they need to applied in any order, they don't conflict and the auto complete works just fine with his patch as is

**Comment by ehelms** [comment]
@jlsherrill thanks for giving it a try in conjunction with this, one last comment (https://github.com/Katello/katello/pull/5294#discussion_r32544142) and I think I am good with this.

File: app/controllers/katello/api/v2/repositories_controller.rb (hunk 2 above)
Comment (jlsherrill) [inline app/controllers/katello/api/v2/repositories_controller.rb]: Changed to use Repository.readable, and updated .readable here:  https://github.com/Katello/katello/pull/5317

**Comment by ehelms** [comment]
ACK

**Pull request merged**